pkg/helper/cache.go

```diff
@@ -0,0 +1,226 @@
+package helper
+
+import (
+	"crypto/md5"
+	"fmt"
+	"io"
+	"reflect"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+)
+
+// Learned from library-go but add generation
+
+type cachedVersionKey struct {
+	name      string
+	namespace string
+	kind      schema.GroupKind
+}
+
+// record of resource metadata used to determine if its safe to return early from an ApplyFoo
+// resourceHash is an ms5 hash of the required in an ApplyFoo that is computed in case the input changes
+// resourceVersion is the received resourceVersion from the apiserver in response to an update that is comparable to the GET
+type cachedResource struct {
+	resourceHash, resourceVersion string
+	generation                    int64
```

Comment:
I think I get it.  If the generation  hasn't changed, then the intent of the object has not changed and therefore you don't need to update it even if the resource version changes.